crates/aggchain-proof-contracts/src/lib.rs

```diff
@@ -1,140 +1,247 @@
+            .block(block_number.into())
+            .await
+            .map_err(Error::LocalExitRootError)?;
+
+        Ok(*response._0)
+    }
+}
+
+#[async_trait::async_trait]
+impl<RpcProvider> L2OutputAtBlockFetcher for AggchainContractsRpcClient<RpcProvider>
+where
+    RpcProvider: alloy::providers::Provider + Send + Sync,
+{
+    fn parse_l2_output_root(json: serde_json::Value) -> Result<L2OutputAtBlock, Error> {
+        fn parse_hash(json: &serde_json::Value, field: &str) -> Result<B256, Error> {
+            let value_str = json
+                .get(field)
+                .ok_or(Error::L2OutputAtBlockValueMissing(field.to_string()))?
+                .as_str()
+                .ok_or(Error::L2OutputAtBlockValueMissing(field.to_string()))?;
+
+            B256::from_str(value_str)
+                .map_err(|e| Error::L2OutputAtBlockInvalidValue(field.to_string(), e))
+        }
+
+        let block_ref = json
+            .get("blockRef")
+            .ok_or(Error::L2OutputAtBlockValueMissing("blockRef".to_string()))?;
+
+        Ok(L2OutputAtBlock {
+            version: parse_hash(&json, "version")?,
+            state_root: *parse_hash(&json, "stateRoot")?,
+            withdrawal_storage_root: *parse_hash(&json, "withdrawalStorageRoot")?,
+            latest_block_hash: *parse_hash(block_ref, "hash")?,
+            output_root: *parse_hash(&json, "outputRoot")?,
+        })
+    }
+
+    async fn get_l2_output_at_block(&self, block_number: u64) -> Result<L2OutputAtBlock, Error> {
+        let params = rpc_params![format!("0x{block_number:x}")];
+        let json: serde_json::Value = self
+            .l2_cl_client
+            .request(&crate::config::default_output_at_block_endpoint(), params)
+            .await
+            .map_err(Error::L2OutputAtBlockRetrievalError)?;
+
+        Self::parse_l2_output_root(json)
+    }
+}
 
-        let l2_client = Arc::new(
-            AlloyProvider::new(
-                l2_rpc_endpoint,
-                prover_alloy::DEFAULT_HTTP_RPC_NODE_INITIAL_BACKOFF_MS,
-                prover_alloy::DEFAULT_HTTP_RPC_NODE_BACKOFF_MAX_RETRIES,
-            )
-            .map_err(Error::ProviderInitializationError)?,
+#[async_trait::async_trait]
+impl<RpcProvider> L1RollupConfigHashFetcher for AggchainContractsRpcClient<RpcProvider>
+where
+    RpcProvider: alloy::providers::Provider + Send + Sync,
+{
+    async fn get_rollup_config_hash(&self) -> Result<Digest, Error> {
+        let response = self
+            .aggchain_fep
+            .chainConfigHash()
+            .call()
+            .await
+            .map_err(Error::RollupConfigHashError)?;
+
+        Ok(*response._0)
+    }
+}
+
+impl AggchainContractsRpcClient<AlloyFillProvider> {
+    pub async fn new(
+        l1_rpc_endpoint: &Url,
+        l2_el_rpc_endpoint: &Url,
+        l2_cl_rpc_endpoint: &Url,
+        network_id: u32,
+        contracts: &AggchainProofContractsConfig,
+    ) -> Result<Self, crate::Error> {
```

Comment:
Does the different RPCs shouldn't be part of the config?

Reply:
@Freyskeyd That was my dilemma. They are already part of the config for the `aggchain-proof-builder`, didn't want to duplicate them in the configuration hence this approach.

Reply:
Since they are actually not directly used by the `aggchain-proof-builder`, I'll remove them from its configuration and put them only in the `AggchainContractsClient` configuration.